Debug assertion hit in `WebCore::FrameLoader::checkLoadCompleteForThisFrame`

rr-codes opened the pull request:
#### 8cf368ed3d211063f591761aaa24bf3d5e4deda3
<pre>
Debug assertion hit in `WebCore::FrameLoader::checkLoadCompleteForThisFrame`
<a href="https://bugs.webkit.org/show_bug.cgi?id=258830">https://bugs.webkit.org/show_bug.cgi?id=258830</a>
rdar://111709160

Reviewed by Wenson Hsieh.

It is prohibited for enum classes that are used in OptionSets to have cases whose value is `0`,
and leads to a debug assertion, because an enum with a value of `0` in an `OptionSet` is
equivalent to an empty OptionSet.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/editing/cocoa/DataDetectorType.h:

Canonical link: <a href="https://commits.webkit.org/265743@main">https://commits.webkit.org/265743@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3e810662c9b0a972e305907e0fde306a684ae4be

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11763 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11962 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12330 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13407 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11203 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11782 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14351 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11943 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14067 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11927 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12762 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/9984 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13828 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10054 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10678 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/10011 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11131 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10833 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14003 "Passed tests") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/11174 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11239 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9282 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/11846 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10414 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3190 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2834 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14697 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/12181 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11096 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2956 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -1964,11 +1964,11 @@ DataDetectorTypes:
   defaultValue:
     WebKitLegacy:
-      default: DataDetectorType::None
+      default: static_cast<DataDetectorType>(0)
```
Would using `refinedType: OptionSet<WebCore::DataDetectorType>` above work instead?

rr-codes replied:
nope :(

```
/Users/richardrobinson/Build/Products/Debug-maccatalyst/usr/local/include/wtf/CrossThreadCopier.h:88:39: error: no member named 'isolatedCopy' in 'WTF::OptionSet<WebCore::DataDetectorType>'
        return std::forward<U>(value).isolatedCopy();
/Users/richardrobinson/Build/Products/Debug-maccatalyst/usr/local/include/wtf/CrossThreadCopier.h:246:55: note: in instantiation of function template specialization 'WTF::CrossThreadCopierBase<false, false, WTF::OptionSet<WebCore::DataDetectorType>>::copy<const WTF::OptionSet<WebCore::DataDetectorType> &>' requested here
    return CrossThreadCopier<std::remove_cvref_t<T>>::copy(std::forward<T>(source));
/Users/richardrobinson/Build/Products/Debug-maccatalyst/DerivedSources/WebCore/Settings.cpp:1923:9: note: in instantiation of function template specialization 'WTF::crossThreadCopy<const WTF::OptionSet<WebCore::DataDetectorType> &>' requested here
        crossThreadCopy(dataDetectorTypes),
1 error generated.
```

darinadler replied:
The default value is an `OptionSet<DateDetectorType>` value, not a `DataDetectorType` value. I suggest trying `{ }`.

rr-codes committed:
Debug assertion hit in `WebCore::FrameLoader::checkLoadCompleteForThisFrame`
https://bugs.webkit.org/show_bug.cgi?id=258830
rdar://111709160

Reviewed by Wenson Hsieh.

It is prohibited for enum classes that are used in OptionSets to have cases whose value is `0`,
and leads to a debug assertion, because an enum with a value of `0` in an `OptionSet` is
equivalent to an empty OptionSet.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/editing/cocoa/DataDetectorType.h:

Canonical link: https://commits.webkit.org/265743@main
